# Problem

**General Dataset Cleaning Instructions**
1.	Duplicate Records: Check for and remove any duplicate rows. Duplicates can skew your analysis and lead to incorrect conclusions.

2.	Check out for Missing Values

3.	Ensure all dates are in a consistent format. 

4.	Standardize the payment type names (e.g., "Credit Card", "Debit Card", "Cash", "Paypal"). This may involve renaming entries to a consistent naming convention.

5.	Negative Values: Look for negative values in columns where it doesn't make sense

6.	Unrealistic Values: Identify outliers in columns like 
 
7.	Typos and Textual Errors:

8.	Correct typos in product names, 

9.	Data Types: Ensure each column is of the right data 
N/B columns that are not involved in calculation should be integers or strings if they do not involve calculations).

10.	Lastly : Correct the entire columns name that are spelt wrongly (TransactionID, Date, CustomerID, ProductID, ProductName, Quantity, Price, TotalSales, PaymentType,  CustomerAge)




***DUE 8PM TUESDAY**

# Solution

In other to start the cleaning process certain python packages must to import from the python libraries which would enable reading, cleaning and writing of our messy data.

1. Numpy - For Numerical Calculations
2. Pandas - For Data Analysis and Manipulation

In [1]:
# Import the Necessary Packages
import numpy as np 
import pandas as pd

Load the messy dataset into a pandas DataFrame named data using the pandas read_csv function because the dataset is in .csv format.

In [2]:
#Load the Messy data into a DataFrame using the read_csv function.
data = pd.read_csv('messy_data.csv')

Lets have a look of the data DataFrame using the pandas .head() function - which shows the first 5 records in the DataFrame.

In [3]:
#View the first five records in the DataFrame
data.head()

TrnsnID        Data  CusmrID  ProdtID ProdtName  Qutity   Prize  TtlSales  \
0     7402  03/17/2026      991     1045     Boo E       9  167.72   2935.06   
1     5835  26-01-2035      741     1040    Book J      18  193.27   4386.52   
2     2123  05/26/2035      933     1062     Bok N       5  126.39   1948.46   
3     8789  28-02-2036      641     1060    Book I       6   66.98   1757.26   
4      305  19-04-2023      123     1058    Book B      15   33.63   2578.19   

  PaymentType  CusterAge  
0         NaN         87  
1  debit card         69  
2        Cash         15  
3        Cash         22  
4        Cash         -1

Lets see the size of the dataset before making an inference, using the .shape method.

In [4]:
data.shape #Shape of the dataset

(10100, 10)

This shows that our dataset comprises of 10100 rows and 10 columns

From above Table view, the first noticeable issue is Column names not spelt correctly, therefore there is a need to rename the columns in order to aid the exploration of the data.

In [5]:
# Rename Columns
old_col_names = (data.columns)
new_col_names = "TransactionID,Date,CustomerID,ProductID,ProductName,Quantity,Price,TotalSales,PaymentType,CustomerAge".split(',')

col_names = {}
for i in range(len(old_col_names)):
    col_names[old_col_names[i]] = new_col_names[i]

data.rename(columns=col_names, inplace=True)
data.head()

TransactionID        Date  CustomerID  ProductID ProductName  Quantity  \
0           7402  03/17/2026         991       1045       Boo E         9   
1           5835  26-01-2035         741       1040      Book J        18   
2           2123  05/26/2035         933       1062       Bok N         5   
3           8789  28-02-2036         641       1060      Book I         6   
4            305  19-04-2023         123       1058      Book B        15   

    Price  TotalSales PaymentType  CustomerAge  
0  167.72     2935.06         NaN           87  
1  193.27     4386.52  debit card           69  
2  126.39     1948.46        Cash           15  
3   66.98     1757.26        Cash           22  
4   33.63     2578.19        Cash           -1

From the Table view above, you will notice that all our column names are now more in a readable form.

The second noticeable issue, is that ProductName and PaymentType values are not standardized across the entire dataset.

Check below for an example:

In [6]:
print(data['ProductName'].value_counts().index)
print(data['PaymentType'].value_counts().index)

Index(['Boo E', 'Book O', 'Book I', 'Book M', 'Book D', 'Book L', 'Book J',
       'Book S', 'Book P', 'Book R', 'Book K', 'Book G', 'Book T', 'Bok N',
       'Bookk H', 'Book F', 'Book A', 'Book B'],
      dtype='object', name='ProductName')
Index(['Debit', 'Credit', 'Cash', 'debit card', 'Credit Card', 'paypal'], dtype='object', name='PaymentType')


The Product Name tells us that this dataset belows to a book selling company - therefore all their sellable products must be a book, but by looking at the above output you will notice that some of the values where not spelt correctly, likewise Payment Type values.

Lets Correct the Payment Type and Product Name Values using the pandas .replace function

In [7]:

data['PaymentType'].replace({'Debit':'Debit Card', 
                             'Credit':'Credit Card', 
                             'debit card':'Debit Card', 
                             'paypal':'Paypal'}, 
                            inplace=True)

data['ProductName'].replace({'Boo E':'Book E', 'Bok N':'Book N', 'Bookk H': 'Book H',}, inplace=True)

data.head()

TransactionID        Date  CustomerID  ProductID ProductName  Quantity  \
0           7402  03/17/2026         991       1045      Book E         9   
1           5835  26-01-2035         741       1040      Book J        18   
2           2123  05/26/2035         933       1062      Book N         5   
3           8789  28-02-2036         641       1060      Book I         6   
4            305  19-04-2023         123       1058      Book B        15   

    Price  TotalSales PaymentType  CustomerAge  
0  167.72     2935.06         NaN           87  
1  193.27     4386.52  Debit Card           69  
2  126.39     1948.46        Cash           15  
3   66.98     1757.26        Cash           22  
4   33.63     2578.19        Cash           -1

Now we easily notice too more issues:

1. PaymentType as a NaN value
2. CustomerAge as a Negative value

But before we can make any changes lets look at the overall position of our dataset.

Let get a brief description of each columns in the dataset

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10100 entries, 0 to 10099
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   TransactionID  10100 non-null  int64  
 1   Date           10100 non-null  object 
 2   CustomerID     10100 non-null  int64  
 3   ProductID      10100 non-null  int64  
 4   ProductName    9098 non-null   object 
 5   Quantity       10100 non-null  int64  
 6   Price          10100 non-null  float64
 7   TotalSales     10100 non-null  float64
 8   PaymentType    8645 non-null   object 
 9   CustomerAge    10100 non-null  int64  
dtypes: float64(2), int64(5), object(3)
memory usage: 789.2+ KB


The output above proved our observation above, it shows that

1. ProductName and PaymentType columns contain missing values
2. Its also shows that we have 7 numerical data types and 3 object or string data type
2. TransactionId, CustomerID, ProductID are simply identifiers, identifying customers and products and transaction history,performing numerical calculations will not identify patterns in customers and products flow. Therefore the columns should be casted to string to enable qualitative analysis be carried-out.

Lets First cast the required columns

In [9]:
#Columns Casting
data['TransactionID'] = data['TransactionID'].astype('str')
data['CustomerID'] = data['CustomerID'].astype('str')
data['ProductID'] = data['ProductID'].astype('str')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10100 entries, 0 to 10099
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   TransactionID  10100 non-null  object 
 1   Date           10100 non-null  object 
 2   CustomerID     10100 non-null  object 
 3   ProductID      10100 non-null  object 
 4   ProductName    9098 non-null   object 
 5   Quantity       10100 non-null  int64  
 6   Price          10100 non-null  float64
 7   TotalSales     10100 non-null  float64
 8   PaymentType    8645 non-null   object 
 9   CustomerAge    10100 non-null  int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 789.2+ KB


Columns string casting completed

Let view a statistically summary of the dataset using the pandas .describe() function.


In [10]:
data.describe().T

count         mean          std    min       25%      50%  \
Quantity     10100.0     7.028119     7.207902  -5.00     1.000     7.00   
Price        10100.0   102.397194    56.429556   5.03    54.095   102.08   
TotalSales   10100.0  2521.484192  1431.601571  20.46  1288.860  2520.87   
CustomerAge  10100.0    44.456832    32.037813 -10.00    17.000    45.00   

                   75%      max  
Quantity       13.0000    19.00  
Price         151.6475   199.98  
TotalSales   3755.7475  4999.31  
CustomerAge    73.0000    99.00

Wow, out if 10 columns only 4 columns are display, why: because the describe function without passing any parameters simply works on only the numerical columns in the DataFrame.

Observation:

1. The Minimum quantity sold = -5, This could probably be a typo error: therefore the mean provided is inaccurate.
2. Customer Age cannot be negative.

Lets deal with the negative values

Lets first Filter the quantity column to get more information

In [11]:
data[(data['Quantity'] < 0)].head()

TransactionID        Date CustomerID ProductID ProductName  Quantity  \
26          4302  12-10-2029        566      1022      Book N        -4   
35          4429  03/28/2036        295      1057      Book E        -4   
47          2376  17-12-2023        243      1075      Book R        -2   
50           316  02/14/2028        566      1012      Book H        -1   
51          8616  06/01/2031        897      1004      Book M        -3   

     Price  TotalSales  PaymentType  CustomerAge  
26  111.77     3955.85  Credit Card           57  
35  137.59     3152.31       Paypal           93  
47  191.13     1900.42  Credit Card          -10  
50  185.04     4526.58  Credit Card           99  
51  151.38       62.12   Debit Card           23

Lets Check for how many rows are affected

In [12]:
data[(data['Quantity'] < 0)].shape

(1960, 10)

1960 rows has negative values, therefore this shows that its a typo error.

Lets now remove the negative sign from all values. 

In [13]:
data['Quantity'] = data['Quantity'].abs()

data[(data['Quantity'] < 0)].shape
    

(0, 10)

We have cleaned the negative value from the Quantity Column.

Lets Look at the Customer Age Column

In [14]:
data[(data['CustomerAge'] <0)]

TransactionID        Date CustomerID ProductID ProductName  Quantity  \
4               305  19-04-2023        123      1058      Book B        15   
6              2996  04/03/2031        571      1086      Book D        15   
30             4443  01/09/2026        294      1088      Book P         6   
32             4534  10/23/2030        963      1052      Book L         3   
47             2376  17-12-2023        243      1075      Book R         2   
...             ...         ...        ...       ...         ...       ...   
10065          7735  10/11/2026        620      1029      Book N         2   
10068          3006  01/23/2025        261      1011      Book I        15   
10075          7850  12/25/2034        678      1005      Book S         5   
10080          3386  04/01/2024        875      1036         NaN        15   
10098           861  04/08/2034        579      1014      Book N        19   

        Price  TotalSales  PaymentType  CustomerAge  
4       33.63     2578.19         Cash           -1  
6      176.27     4822.26         Cash           -4  
30     137.01     2783.85          NaN           -8  
32     152.39      761.66         Cash           -1  
47     191.13     1900.42  Credit Card          -10  
...       ...         ...          ...          ...  
10065   50.81     1780.35  Credit Card           -5  
10068  192.64     4645.92   Debit Card           -4  
10075  102.58     4466.09       Paypal           -7  
10080   70.55     4653.93   Debit Card           -1  
10098  137.58     1708.87   Debit Card           -7  

[985 rows x 10 columns]

We can see that we have almost a 1000 points that shows Age with negative value, but with this alone, suggestion on how to deal with the negative value cannot be made.

Lets check for more:

In [15]:
data[(data['CustomerAge'] >0) & (data['CustomerAge'] <10)]

TransactionID        Date CustomerID ProductID ProductName  Quantity  \
15             6709  08/20/2029        583      1044      Book T         8   
38             7788  02/03/2027        382      1059      Book R         2   
59             8569  27-05-2027        423      1097      Book L        12   
63             6740  28-04-2029        871      1088      Book J         9   
81             5401  08/05/2030        953      1095      Book T        14   
...             ...         ...        ...       ...         ...       ...   
10074          2048  04-08-2029        637      1022      Book I         3   
10083          6421  24-07-2032        685      1036      Book G        11   
10084          5052  06-07-2023         90      1036      Book J         5   
10090          8323  15-03-2033        157      1090      Book E         1   
10091          5579  10/28/2024        276      1007      Book S        15   

        Price  TotalSales  PaymentType  CustomerAge  
15     108.44     2664.90         Cash            9  
38     101.96     2475.43         Cash            6  
59     112.45      485.15  Credit Card            8  
63      27.93     3422.45   Debit Card            4  
81     118.37     4176.75       Paypal            7  
...       ...         ...          ...          ...  
10074  106.15     2330.17   Debit Card            2  
10083  146.44     1216.02   Debit Card            3  
10084  116.58     1191.14  Credit Card            7  
10090   76.06     3288.01   Debit Card            3  
10091   78.54     3526.21   Debit Card            4  

[793 rows x 10 columns]

In [16]:
data[(data['CustomerAge'] >90)]

TransactionID        Date CustomerID ProductID ProductName  Quantity  \
9              4449  02/05/2030        614      1086      Book M         5   
17             3320  08/16/2032        180      1059      Book G        17   
35             4429  03/28/2036        295      1057      Book E         4   
50              316  02/14/2028        566      1012      Book H         1   
52             3527  02-02-2026        926      1051      Book B         3   
...             ...         ...        ...       ...         ...       ...   
10058          8793  02/04/2036        487      1067      Book B        19   
10060          8839  29-10-2027        865      1008      Book N        19   
10062          3557  21-06-2034        757      1053      Book I         1   
10079          6397  04/12/2025        469      1030      Book O        14   
10095          5735  04/03/2027         52      1066      Book G         1   

        Price  TotalSales  PaymentType  CustomerAge  
9       15.67      714.58   Debit Card           95  
17      15.92      861.47  Credit Card           97  
35     137.59     3152.31       Paypal           93  
50     185.04     4526.58  Credit Card           99  
52      93.77     2805.37   Debit Card           91  
...       ...         ...          ...          ...  
10058   23.24     3475.14          NaN           91  
10060   60.21      351.13   Debit Card           99  
10062   22.26     4975.04  Credit Card           99  
10079  141.10     3604.93  Credit Card           94  
10095  191.41     2788.25  Credit Card           91  

[838 rows x 10 columns]

Data Integrity Issues:

Firstly the product for sale are books, then why does the data include Customers of Ages less than 10 Years and Ages greater than 90 years.

It also include data at the extreme points of 1year and 99 years.

So therefore the solution for the negative value will simple be convert negative to positive, then looking into the data source for futher corrections.

In [17]:
data['CustomerAge'] = data['CustomerAge'].abs()

data[(data['CustomerAge'] < 0)].shape
    

(0, 10)

We Can now see that none of the Customer Age is negative

Now Lets move to the String or Object data types


In [18]:
data.describe(include='object').T

count unique         top  freq
TransactionID  10100  10000        1964     2
Date           10100  10000  05/27/2027     2
CustomerID     10100    999         794    20
ProductID      10100    100        1053   124
ProductName     9098     18      Book E   574
PaymentType     8645      4  Debit Card  2901

With include='object' has parameter in the describe function we can get a statistical summary of the objects columns

Observations:

1. TransactionId has only 10000 unique values, that means there is a duplicate of 100 rows, because every transaction must be unique even if made by the same customer.
2. The Company has 999 customers and there Top customer is 794 with 20 purchases.
3. The Company has only 100 productID, 18 ProductName and only 4 types of Payment methods.

Now Before we perform further analysis lets remove the duplicates

In [19]:
data[data['TransactionID'].duplicated()].shape

(100, 10)

The above confirms our observations.

Now lets remove removes the duplicates using the TransactionID column, because each Transaction must be Unique.

In [20]:
data.drop_duplicates(subset='TransactionID', inplace=True)

data.shape

(10000, 10)

All 100 Duplicated Rows have been dropped.

Now lets work on the Missing Values

In [21]:
#Find Missing Values
data.isna().sum()

TransactionID       0
Date                0
CustomerID          0
ProductID           0
ProductName       994
Quantity            0
Price               0
TotalSales          0
PaymentType      1439
CustomerAge         0
dtype: int64

In [22]:
print(data['ProductName'].value_counts())

print(data['PaymentType'].value_counts())

ProductName
Book E    565
Book O    547
Book I    523
Book M    522
Book L    513
Book D    510
Book J    504
Book R    500
Book P    500
Book S    499
Book K    495
Book T    492
Book G    489
Book H    489
Book N    486
Book F    480
Book A    448
Book B    444
Name: count, dtype: int64
PaymentType
Debit Card     2869
Credit Card    2852
Cash           1426
Paypal         1414
Name: count, dtype: int64


From Observations done from various permutations, we saw using the must frequent variable to fill the missing values might skew the dataset, by overshooting the value leading to false data.abs

Due to Knowledge Constraint we will be using the ffill method to fill the missing values.

In [23]:
data = data.fillna(method='ffill')
data['PaymentType'] = data['PaymentType'].fillna(method='bfill')

In [24]:
print(data['ProductName'].value_counts())
print(data['PaymentType'].value_counts())


ProductName
Book E    623
Book O    616
Book M    582
Book D    571
Book J    567
Book I    566
Book L    563
Book P    560
Book R    559
Book K    557
Book S    556
Book T    544
Book G    540
Book N    537
Book H    536
Book F    528
Book A    500
Book B    495
Name: count, dtype: int64
PaymentType
Debit Card     3347
Credit Card    3343
Cash           1674
Paypal         1636
Name: count, dtype: int64


In [25]:
print(data.isna().sum())

TransactionID    0
Date             0
CustomerID       0
ProductID        0
ProductName      0
Quantity         0
Price            0
TotalSales       0
PaymentType      0
CustomerAge      0
dtype: int64


We can now see that how data points are almost even increased based, rather than increasing only the most frequent value in the Product Name and Payment Type column

Now Lets clean up the Data column

We would be using the pandas datetime function for ease

In [26]:
data['Date'] = pd.to_datetime(data['Date'], format='mixed')
data['Date'].head()

0   2026-03-17
1   2035-01-26
2   2035-05-26
3   2036-02-28
4   2023-04-19
Name: Date, dtype: datetime64[ns]

In [27]:
data.sample(10)

TransactionID       Date CustomerID ProductID ProductName  Quantity  \
99            4831 2029-03-10        376      1015      Book J         5   
2503          1615 2030-11-19        105      1059      Book D         4   
2078          9053 2025-02-18        152      1058      Book R         5   
5134          6731 2028-12-12         84      1035      Book I         0   
4144          3160 2030-12-23        462      1031      Book G         1   
2398           248 2031-05-29        461      1015      Book E         5   
5811          3752 2024-01-14        843      1051      Book E         3   
5850          4397 2035-10-13        476      1044      Book H         2   
8398          6472 2032-04-18        771      1070      Book A        18   
3696          1488 2027-03-24        665      1045      Book H         3   

       Price  TotalSales  PaymentType  CustomerAge  
99     16.30     1169.97  Credit Card           10  
2503  128.18     4294.32   Debit Card           66  
2078  111.02     4698.67  Credit Card           63  
5134  179.60     3569.25       Paypal            2  
4144   60.15     1738.70   Debit Card           90  
2398   46.19      675.69   Debit Card            6  
5811  183.32     3964.50  Credit Card            8  
5850  147.58      974.74   Debit Card           40  
8398  101.80     1686.91         Cash           38  
3696   29.06     4434.81   Debit Card           87

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10000 entries, 0 to 10099
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   TransactionID  10000 non-null  object        
 1   Date           10000 non-null  datetime64[ns]
 2   CustomerID     10000 non-null  object        
 3   ProductID      10000 non-null  object        
 4   ProductName    10000 non-null  object        
 5   Quantity       10000 non-null  int64         
 6   Price          10000 non-null  float64       
 7   TotalSales     10000 non-null  float64       
 8   PaymentType    10000 non-null  object        
 9   CustomerAge    10000 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(2), object(5)
memory usage: 859.4+ KB


The Data has been successfully cleaned.

## Summary

1. The Data know comprises of 10000 rows and 10 columns, where we have 5 object columns, 4 numerical columns and a datetime column
2. No Missing Value
3. No Duplicates
4. All values and headers have been Standardized


### Data Integrity Issues:

1. Giving that this is a book selling company, there should not be customers below 10 years old buying books.
2. We could also see data points of age above 90 years, what are they reading?
3. The Total sales column does not match the basic numerical computation of Price X Quantity. 

More understanding about the source need to be carried out in other to solve the above integrity issues


### Constraint: Knowledge Gap


Based on the distribution of the Product Name and Payment Type a more robust method for filling the missing value should have been used, e.g

1. Product Name should be filled based on the most popular product filtered based on age range (Age 1 to 15 likes this type of books, e.t.c)
2. Payment Type missing value should be filled based on customers spending habit or culture.





In [29]:
# Export Clean Data

data.to_csv('clean_data.csv', index='False')